lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,953 @@
+			case *completeType:
+				select {
+				case completeChan <- rType.RegistrationComplete:
+				case <-returnedChan:
+				}
+
+			// An error occurred during the registration process.
+			case *walletrpc.RegistrationResponse_RegistrationError:
+				err := fmt.Errorf("registration error: %s",
+					rType.RegistrationError)
+
+				select {
+				case errChan <- err:
+				case <-returnedChan:
+				}
+			}
+
+			return
+
+		default:
+			err := fmt.Errorf("expected registration response, "+
+				"but got: %T", reqType)
+
+			select {
+			case errChan <- err:
+			case <-returnedChan:
+			}
+
+			return
+		}
+	})
+	if !success {
+		return errors.New("error starting registration completion " +
+			"checking function")
+	}
+
+	// Wait for the watch-only node to respond that it has accepted the
+	// signer has registered.
+	select {
+	case <-completeChan:
+		// TODO(viktor): This should verify that the signature in the
+		// complete message is valid.
+
+	case err := <-errChan:
+		return fmt.Errorf("watch-only node handshake error: %w", err)
+
+	case <-ctxt.Done():
+		return ctxt.Err()
+	}
+
+	return nil
+}
+
+// processSignRequests processes and responds to the sign requests that are
+// sent over the stream. The function will continuously run until the remote
+// signer client is either stopped or the stream errors.
+func (r *OutboundClient) processSignRequests(ctx context.Context,
+	stream *Stream) error {
+
+	var (
+		reqChan = make(chan *walletrpc.SignCoordinatorRequest)
+		errChan = make(chan error)
+	)
+	// We don't defer closing the channels here because doing so could
+	// cause a race condition in the goroutine below, where it might attempt
+	// to send on a closed channel, leading to a panic. Instead, the garbage
+	// collector will clean up the channels once this function returns and
+	// the goroutine has completed execution.
+
+	// We run the receive loop in a goroutine to ensure we can stop if the
+	// remote signer client is shutting down (i.e. the quit channel is
+	// closed). Closing the quit channel will make the processSignRequests
+	// function return, which will cancel the stream context, which in turn
+	// will stop the receive goroutine.
+	success := r.gManager.Go(func(_ context.Context) {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				wrappedErr := fmt.Errorf("error receiving "+
+					"request from watch-only node: %w", err)
+
+				// Send the error to the error channel, given
+				// that we're still listening on the channel.
+				select {
+				case errChan <- wrappedErr:
+				case <-ctx.Done():
+				}
+
+				return
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+
+			case reqChan <- req:
+			}
+		}
+	})
+	if !success {
+		return errors.New("error starting receiving loop")
+	}
+
+	for {
+		log.Tracef("Waiting for a request from the watch-only node")
+
+		select {
+		case req := <-reqChan:
+			// Process the received request.
+			err := r.handleRequest(ctx, req, stream)
```

Comment:
in general, this pattern of chaining calls is not great andis hard to read. It's easier to read something that goes:

1) get request
2) process request
3) form response
4) respond

instead of each one calling into the next one. 

See the patch i posted for this :)

Reply:
Thanks a lot, totally agree that this is more readable :tada:!